internal/cmd/io_binary.go

```diff
@@ -0,0 +1,260 @@
+
+	binaryCloser sync.Once
+
+	stdin, stdout, stderr string
+
+	sout, serr, wait io.ReadWriteCloser
+	soutCloser       sync.Once
+}
+
+// NewBinaryIO starts a binary logger
+func NewBinaryIO(ctx context.Context, id string, uri *url.URL) (_ UpstreamIO, err error) {
+	ns, err := namespaces.NamespaceRequired(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	var (
+		sout, serr, w io.ReadWriteCloser = nil, nil, nil
+	)
+
+	bio := &binaryIO{}
+
+	stdoutPipe := fmt.Sprintf(`%s\binary-%s-stdout`, pipeRootBinary, id)
+	sout, err = openNPipe(stdoutPipe)
+	if err != nil {
+		return nil, err
+	}
+	bio.stdout = stdoutPipe
+	bio.sout = sout
+
+	stderrPipe := fmt.Sprintf(`%s\binary-%s-stderr`, pipeRootBinary, id)
+	serr, err = openNPipe(stderrPipe)
+	if err != nil {
+		return nil, err
+	}
+	bio.stderr = stderrPipe
+	bio.serr = serr
+
+	waitPipe := fmt.Sprintf(`%s\binary-%s-wait`, pipeRootBinary, id)
+	w, err = openNPipe(waitPipe)
+	if err != nil {
+		return nil, err
+	}
+	bio.wait = w
+
+	cmd := newBinaryCmd(ctx, uri, id, ns)
+	cmd.Env = append(cmd.Env,
+		"CONTAINER_STDOUT="+stdoutPipe,
+		"CONTAINER_STDERR="+stderrPipe,
+		"CONTAINER_WAIT="+waitPipe,
+	)
+
+	bio.cmd = cmd
+
+	started := make(chan bool, 1)
+	if err := cmd.Start(); err != nil {
+		return nil, err
+	}
+
+	go func() {
+		defer w.Close()
+		t := 0
+		for {
+			b := make([]byte, 1)
+			v, err := w.Read(b)
+			if (v == 0) || err != nil && err != io.EOF {
+				log.G(ctx).Debugf("Failed to read from wait pipe. Sleeping")
+				time.Sleep(1 * time.Second)
+				t++
+				if t > 10 {
+					break
+				}
+			} else {
+				log.G(ctx).Debugf("Read from wait pipe. Binary started: %s", b)
+				started <- true
+				return
+			}
+		}
+		started <- false
+	}()
+
+	if !<-started {
+		return nil, errors.Errorf("Failed to started binary")
+	}
+
+	return bio, nil
+}
+
+var _ UpstreamIO = &binaryIO{}
+
+func (b *binaryIO) Close(ctx context.Context) {
+	b.soutCloser.Do(func() {
+		if b.sout != nil {
+			err := b.sout.Close()
+			if err != nil {
+				log.G(ctx).WithError(err).Errorf("Error while closing stdout npipe")
+			}
+		}
+		if b.serr != nil {
+			err := b.serr.Close()
+			if err != nil {
+				log.G(ctx).WithError(err).Errorf("Error while closing stderr npipe")
+			}
+		}
+	})
+	b.binaryCloser.Do(func() {
+		// Borrowed this from Ming's PR
+		log.G(ctx).Debugf("Waiting for binaryIO to exit: %d", b.cmd.Process.Pid)
+		done := make(chan error, 1)
+		go func() {
+			done <- b.cmd.Wait()
+		}()
+
+		select {
+		case err := <-done:
+			if err != nil {
+				log.G(ctx).WithError(err).Errorf("Error while waiting for cmd to finish")
+			} else {
+				log.G(ctx).Debugf("binary_io::b.cmd.Wait() finished normally")
+			}
+
+		case <-time.After(binaryCmdWaitTimeout):
+			log.G(ctx).Errorf("Timeout while waiting for binaryIO process to finish")
+			err := b.cmd.Process.Kill()
+			if err != nil {
+				log.G(ctx).WithError(err).Errorf("Error while killing binaryIO process")
+			}
+			log.G(ctx).Debugln("BinaryIO process killed")
```

Comment:
We'll end up printing this regardless of if it was actually killed as we don't return even if the above errored out

Reply:
yeah, I'm gonna cleanup the success branches